images/capi/ansible/roles/containerd/tasks/main.yml

```
@@ -39,6 +39,45 @@
     dest: /
     extra_opts:
       - --no-overwrite-dir
+  when: ansible_os_family != "Flatcar"
+
+- name: unpack containerd for Flatcar into /opt
+  unarchive:
+    remote_src: True
+    src: /tmp/containerd.tar.gz
+    dest: /opt
+    exclude:
+      - ./usr/local/bin/containerd
```

Is Flatcar running a customized/forked version of containerd? Should we not be overwriting the built-in version with the user-specified version to make the containerd version consistent across builds ?

---

> Is Flatcar running a customized/forked version of containerd? Should we not be overwriting the built-in version with the user-specified version to make the containerd version consistent across builds ?

Flatcar runs a vanilla containerd, 1.1 for stable and 1.3 for beta/alpha as of now. However, containerd in Flatcar is one of the packages protected by [torcx](https://github.com/flatcar-linux/torcx), a boot-time add-on manager. It means, its built-in binary `/run/torcx/unpack/docker/bin/containerd` is located under a read-only partition `/run/torcx/unpack`, and the mount point is "sealed" through the entire machine's lifetime. So it is not possible overwrite the build-in containerd binary with another binary.

The downloaded binary `/usr/local/bin/containerd` will be kept separately from the torcx-managed binaries.

---

Perhaps I am missing something, but if you are excluding the containerd binaries, then why download the containerd archive at all?  I don't see where `/opt/bin/containerd` is created our used (The systemd unit still uses `${TORCX_BINDIR}/containerd` )

---

> Perhaps I am missing something, but if you are excluding the containerd binaries, then why download the containerd archive at all? I don't see where /opt/bin/containerd is created our used (The systemd unit still uses ${TORCX_BINDIR}/containerd )


Let me go through the whole process in `ansible/roles/containerd/tasks/main.yml`.

1. download containerd
 : It downloads a tarball from [containerd_url](https://storage.googleapis.com/cri-containerd-release/cri-containerd-1.3.4.linux-amd64.tar.gz), and stores it into `/tmp/containerd.tar.gz`. The tarball includes all the containerd binaries to be unpacked into the target filesystem, including `/opt`, '/etc`, and `/usr`.

2. unpack containerd into /opt
 : It unpacks `/tmp/containerd.tar.gz` into `/opt`, excluding binaries like `/usr/local/bin/containerd` or `/usr/local/sbin/runc`, because they are already provided by Flatcar itself.
 
   On the other hand, it still needs to install other binaries like `/usr/local/bin/containerd-shim` or `/usr/local/bin/crictl`, as they are not included in Flatcar. Obviously it has to install the binaries into the target directory `/opt`, because `/usr` in Flatcar is read-only.

3. unpack containerd into /etc
 : unpacks `/tmp/containerd.tar.gz` into `/etc`, excluding other binaries as well as some unnecessary config files.

We need to go through multiple steps that might look complicated, because we need to install only necessary binaries that are included in the downloaded cri-containerd tarball, without trying to touch anything in read-only partitions.

---

| unpack containerd into /opt

The issue is that we are not unpacking containerd, rather some other bits that are not containerd which breaks an implicit contract: If I provide a specific containerd version to image-builder, I expect that version to be used in the resulting image (I understand *why* you can't use a specific version).

With this patch the installation comprises of a patched containerd binaries supplied by the OS, and some other files supplied by a user-specified containerd version. 

From what I can see in https://github.com/flatcar-linux/coreos-overlay/blob/main/app-emulation/containerd/containerd-9999.ebuild#L61 containerd-shim is already installed?

My suggestions are:

1) If the user-supplied containerd version does not match the containerd version supplied by Flatcar, fail hard
2) Install crictl via https://github.com/kubernetes-sigs/cri-tools
3) Embed and install the containerd config files / units in the playbook

---

> From what I can see in https://github.com/flatcar-linux/coreos-overlay/blob/main/app-emulation/containerd/containerd-9999.ebuild#L61 containerd-shim is already installed?

Only `containerd-shim` is available via torcx.
Neither `containerd-shim-v1` nor `containerd-shim-v2`.
Without installing both containerd shims into `/opt`, v1 and v2, the image-builder simply does not work.
No matter what, we cannot get rid of unpacking the cri-containerd tarball.

> 1. If the user-supplied containerd version does not match the containerd version supplied by Flatcar, fail hard

It makes sense. I will see what I can do.

> 2. Install crictl via https://github.com/kubernetes-sigs/cri-tools

Do you mean, like downloading a specific tarball like https://github.com/kubernetes-sigs/cri-tools/releases/download/v1.16.1/crictl-v1.16.1-linux-amd64.tar.gz ?
I can do that if you want.
Though I am not sure if that's in the scope of this PR.
In the master branch it already downloads the cri-containerd tarball, which includes its crictl binary.
If you mean creating another PR based on master to make it depend on the upstream cri-tools repo, then I can do that as well.

> 3. Embed and install the containerd config files / units in the playbook

I thought we are already doing that.

---

Can `containerd-shim-v1` and `containerd-shim-v2` not be fixed in upstream emulation-package?

2 & 3 are instead of downloading the containterd tarball, not in addition